service/matching/matching_engine.go

```diff
@@ -1212,6 +1219,160 @@ func (e *matchingEngineImpl) ReplicateTaskQueueUserData(ctx context.Context, req
+		return nil, serviceerror.NewFailedPrecondition("versioning is disabled")
+	}
+
+	taskID := uuid.New()
+	resp, err := tqm.DispatchNexusTask(ctx, taskID, request)
+
+	// if we get a response or error it means that the Nexus task was handled by forwarding to another matching host
+	// this remote host's result can be returned directly
+	if resp != nil || err != nil {
+		return resp, err
+	}
+
+	// if we get here it means that dispatch of query task has occurred locally
+	// must wait on result channel to get query result
+	resultCh := make(chan *nexusResult, 1)
+	e.lockableNexusResultMap.put(taskID, resultCh)
+	defer e.lockableNexusResultMap.delete(taskID)
+
+	select {
+	case result := <-resultCh:
+		if result.internalError != nil {
+			return nil, result.internalError
+		}
+		if result.failedWorkerResponse != nil {
+			return &matchingservice.DispatchNexusTaskResponse{Outcome: &matchingservice.DispatchNexusTaskResponse_HandlerError{
+				HandlerError: result.failedWorkerResponse.GetRequest().GetError(),
+			}}, nil
+		}
+
+		return &matchingservice.DispatchNexusTaskResponse{Outcome: &matchingservice.DispatchNexusTaskResponse_Response{
+			Response: result.successfulWorkerResponse.GetRequest().GetResponse(),
+		}}, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (e *matchingEngineImpl) PollNexusTaskQueue(
+	ctx context.Context,
+	req *matchingservice.PollNexusTaskQueueRequest,
+	opMetrics metrics.Handler,
+) (*matchingservice.PollNexusTaskQueueResponse, error) {
+	namespaceID := namespace.ID(req.GetNamespaceId())
+	pollerID := req.GetPollerId()
+	request := req.Request
+	taskQueueName := request.TaskQueue.GetName()
+	e.logger.Debug("Received PollNexusTaskQueue for taskQueue", tag.Name(taskQueueName))
+pollLoop:
+	for {
+		err := common.IsValidContext(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		taskQueue, err := newTaskQueueID(namespaceID, taskQueueName, enumspb.TASK_QUEUE_TYPE_NEXUS)
+		if err != nil {
+			return nil, err
+		}
+
+		// Add frontend generated pollerID to context so taskqueueMgr can support cancellation of
+		// long-poll when frontend calls CancelOutstandingPoll API
+		pollerCtx := context.WithValue(ctx, pollerIDKey, pollerID)
+		pollerCtx = context.WithValue(pollerCtx, identityKey, request.GetIdentity())
+		pollMetadata := &pollMetadata{
+			workerVersionCapabilities: request.WorkerVersionCapabilities,
+		}
+		task, err := e.getTask(pollerCtx, taskQueue, normalStickyInfo, pollMetadata)
+		if err != nil {
+			if err == errNoTasks {
+				return &matchingservice.PollNexusTaskQueueResponse{}, nil
+			}
+			return nil, err
+		}
+
+		e.emitForwardedSourceStats(opMetrics, task.isForwarded(), req.GetForwardedSource())
+
+		if task.isStarted() {
+			// tasks received from remote are already started. So, simply forward the response
+			return task.pollNexusTaskQueueResponse(), nil
+		}
+
+		task.finish(err)
+		if err != nil {
+			continue pollLoop
+		}
+
+		taskToken := &tokenspb.QueryTask{
```

Comment:
I think it's a little confusing to reuse the `QueryTask` type here. Did you consider making the name more generic or adding a `NexusTask`?

Reply:
Yeah, good point, thanks!